---
title: "filtR: a function for filtering error-derived OTUs/ASVs from count tables"
date: July 3rd, 2018
authors: "Ben Joris, Greg Gloor, Jean Macklaim, Daniel Giguere, Kaitlyn Hobbs"
output:
  html_document:
    toc: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteIndexEntry{filtR: a function for filtering error-derived OTUs/ASVs from count tables}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# What is filtR?

`filtR` is an R function with the purpose of filtering out error-derived operational taxonomic units (OTUs) or amplicon sequence variants (ASVs) from count tables. `filtR` integrates the functionality of two R packages, `ALDEx2`[^1] and `propr`[^2]. If an OTU/ASV is deemed to be error-derived by filtR, it will be removed from the count table.

[^1]: Fernandes et al (2013) PLOS ONE https://doi.org/10.1371/journal.pone.0067019
[^2]: Quinn et al (2017) Sci Rep. https://dx.doi.org/10.1038%2Fs41598-017-16520-0

# Introduction

The $\rho$ statistic can be used to assess $\beta$-association between a pair of features [^3]. As the value of $\rho$ approaches 1, it can be assumed that the two features used to calculate $\rho$ have a constant ratio across the dataset and are therefore associated[^4]. The premise of ASVs and OTUs is generally that each individual ASV and OTU, for the most part, should lack association with other ASVs/OTUs because they are independent entities. However, simply a strong association between two features--measured by $\rho$--is not sufficient to identify an ASV/OTU that was generated by errors introduced in the sequencing process. As shown in a study of the microbiota of Chinese individuals[^5], the error-derived OTUs tend to have a difference in centered log ratios (base 2) of greater than 5 compared to their parent OTU. In essence, the rarer OTUs that were strongly associated with the prevalent OTU were likely sequences derived from errors generated during sequencing or PCR. Since these OTUs do not represent real biological diversity, labelling them as their own entity is a false positive when assessing overall diversity of a microbiome.

# Installation and usage

Install the latest version of `filtR` from GitHub with `devtools::install_github('bjoris33/filtR')`.

Load in the R package.
```{r}
library(filtR)
```
With these two steps completed, we are now ready to use the `filtR()` function of the `filtR` package.

# Intial data input

The format of the data that `filtR` requires is a count table with OTUs/ASVs as its rows and samples as the columns. The count table can contain a column for taxonomic assignment, but it is not a requirement. For the purposes of the this vignette, a [count table with taxonomic assignments](https://www.ebi.ac.uk/metagenomics/beta/studies/MGYS00002165) from the [European Bioinformatics Institute (EBI)](https://www.ebi.ac.uk/metagenomics/) will be used as an example.

Let's begin by loading in the count table from EBI:
```{r}
unfilt <- read.table('SRP078943_taxonomy_abundances_SSU_v4.1.tsv')
dim(unfilt)
```
The initial count table has 319 rows (which represent the OTUs) and 38 columns, which are the 37 samples and one column for the taxonomic assignments for the OTUs.

# Applying the filtR function
With a count table selected, we can now use the filtR function:
```{r}
filtered <- filtR('SRP078943_taxonomy_abundances_SSU_v4.1.tsv')
dim(filtered)
```
As we can see from the output, filtR filtered 4 OTUs from the count table. The original dimensions of the count table were 319 rows and 38 columns. Now that the 4 OTUs are filtered, the dimensions are 315 rows and 38 columns. The count table is in the same format it originally was in, just now with the error-derived OTUs removed.

# Modifiable parameters in filtR
`filtR` is a simple function with 3 modifiable parameters:

* `count_file`: the file name of the count table file that is being filtered.

* `rho_CO`: Defaults to 0.7. The value of $\rho$ (range from 0 to 1) that is being used as a cutoff for analysis. Pairs of features with a greater value of $\rho$ will be indexed. If filtR is failing to index any pairs of OTUs/ASVs, this parameter can be lowered. However, this increases the chance that the filter will eliminate 'real' OTUs/ASVs and it is recommended to use the default cutoff or higher.

* `clr_CO`: Defaults to 5. This filtering parameter sets the minimum difference in centered log ratio between the pairs of OTUs indexed by the $\rho$ cutoff. To increase the number of OTUs/ASVs that are filtered by filtR, this parameter can be lowered. However, this increases the chance that the filter will eliminate 'real' OTUs/ASVs and it is recommended to use the default cutoff or higher.

Let's explore how modifying these filtering parameters individually can change the results of filtering.

If we lower the $\rho$ cutoff to 0.6:
```{r}
filtered <- filtR('SRP078943_taxonomy_abundances_SSU_v4.1.tsv', rho_CO = 0.6)
```
Now we have 7 OTUs being filtered from the count table.

If we lower the centered log ratio difference cutoff to 4 from 5 (and keep the $\rho$ cutoff at the default):
```{r}
filtered <- filtR('SRP078943_taxonomy_abundances_SSU_v4.1.tsv', clr_CO = 4)
```
With the CLR difference cutoff lowered, 6 OTUs are filtered from the count table.

# Error handling in filtR

If the $\rho$ cutoff fails to index any pairs, the following error will be generated:

```{r, error = TRUE, results = FALSE}
filtered <- filtR('SRP078943_taxonomy_abundances_SSU_v4.1.tsv', rho_CO = 0.99)
```

If the CLR difference cutoff is larger than the differences between the indexed pairs, the following error will be generated:

```{r, error = TRUE, results = FALSE}
filtered <- filtR('SRP078943_taxonomy_abundances_SSU_v4.1.tsv', clr_CO = 50)
```




[^3]: Erb et al (2016) Theory Biosci. https://doi.org/10.1007/s12064-015-0220-8
[^4]: Lovell et al (2015) Plos Comput Biol. https://doi.org/10.1371/journal.pcbi.1004075
[^5]: Bian et al (2017) mSphere https://doi.org/10.1128/mSphere.00327-17
